s1kd-aspp/doc/s1kd-aspp.1
.\" Automatically generated by Pandoc 1.19.2.1
.\"
.TH "s1kd\-aspp" "1" "2018\-05\-14" "" "s1kd\-tools"
.hy
.SH NAME
.PP
s1kd\-aspp \- Applicability statement preprocessor
.SH SYNOPSIS
.IP
.nf
\f[C]
s1kd\-aspp\ [\-g\ [\-A\ <ACT>]...\ [\-C\ <CCT>]...\ [\-G\ <XSL>]]
\ \ \ \ \ \ \ \ \ \ [\-p\ [\-a\ <ID>]]\ [\-dfxh?]\ [<modules>...]
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[I]s1kd\-aspp\f[] tool has two main functions:
.IP \[bu] 2
Generates display text for applicability statements.
The text is derived from the logic described by the \f[C]assert\f[] and
\f[C]evaluate\f[] elements.
.IP \[bu] 2
Preprocesses "semantic" applicability statements in a data module to
produce "presentation" applicability statements which are simpler to
parse in an XSLT stylesheet.
.RS 2
.PP
"Semantic" applicability statements are those entered by the author to
encode the applicability of elements within a data module.
"Presentation" applicability statements are those that are actually
displayed in page\-oriented output, also referred to as the
"human\-readable" statements.
.PP
The applicability in the resulting XML is longer semantically correct,
but an XSLT stylesheet can simply place a statement on any element with
attribute \f[C]applicRefId\f[] without needing to consider inherited
applicability statements on elements without the attribute.
.RE
.SH OPTIONS
.TP
.B \-A <ACT>
Add an ACT to use when generating display text for product attributes.
Multiple ACT data modules can be used by specifying this option multiple
times.
.RS
.RE
.TP
.B \-a <ID>
The ID to use for the inline applicability annotation representing the
whole data module\[aq]s applicability.
Default is "applic\-0000".
.RS
.RE
.TP
.B \-C <CCT>
Add a CCT to use when generating display text for conditions.
Multiple CCT data modules can be used by specifying this option multiple
times.
.RS
.RE
.TP
.B \-d
Dump the built\-in XSLT used to generate display text for applicability
statements.
.RS
.RE
.TP
.B \-f
Overwrite input data module(s) rather than outputting to stdout.
.RS
.RE
.TP
.B \-G <XSLT>
Use custom XSLT to generate display text for applicability statements.
.RS
.RE
.TP
.B \-g
Generate display text for applicability statements.
.RS
.RE
.TP
.B \-p
Preprocess applicability statements to produce "presentation"
applicability statements which are simpler to parse in an XSLT
stylesheet.
The applicability in the resulting XML is no longer semantically
correct.
.RS
.RE
.TP
.B \-x
Process the modules using the XInclude specification.
.RS
.RE
.TP
.B <modules>...
The module(s) to preprocess.
This can include both individual modules and combined files such as
those produced by s1kd\-flatten(1).
.RS
.RE
.SH EXAMPLES
.SS Generating display text
.PP
The built\-in XSLT for generating display text follows the guidance in
Chap 7.8 of the S1000D 4.2 specification.
For example, given the following:
.IP
.nf
\f[C]
<applic>
<assert\ applicPropertyIdent="prodversion"\ applicPropertyType="prodattr"
applicPropertyValues="A"/>
</applic>
\f[]
.fi
.PP
The resulting XML would contain:
.IP
.nf
\f[C]
<applic>
<displayText>
<simplePara>prodversion:\ A</simplePara>
</displayText>
<assert\ applicPropertyIdent="prodversion"\ applicPropertyType="prodattr"
applicPropertyValues="A"/>
</applic>
\f[]
.fi
.PP
If ACTs or CCTs are supplied which define display names for a property,
this will be used instead of the ident.
For example, the ACT defines the display name for the
"\f[C]prodversion\f[]" product attribute:
.IP
.nf
\f[C]
<productAttribute\ id="prodversion">
<name>Product\ version</name>
<displayName>Version</displayName>
<descr>The\ version\ of\ the\ product.</descr>
<enumeration\ applicPropertyValues="A|B|C"/>
</productAttribute>
\f[]
.fi
.PP
When supplied with the \-A option:
.IP
.nf
\f[C]
$\ s1kd\-aspp\ \-g\ \-A\ <ACT>\ <DM>
\f[]
.fi
.PP
The resulting XML would instead contain:
.IP
.nf
\f[C]
<applic>
<displayText>
<simplePara>Version:\ A</simplePara>
<assert\ applicPropertyIdent="prodversion"\ applicPropertyType="prodattr"
applicPropertyValues="A"/>
</displayText>
</applic>
\f[]
.fi
.PP
The methods for generating display text can be changed by supplying a
custom XSLT script with the \-G option.
The \-d option can be used to dump the built\-in XSLT as a starting
point for a custom script.
An identity template is automatically added to the script, equivalent to
the following:
.IP
.nf
\f[C]
<xsl:template\ match="\@*|node()">
<xsl:copy>
<xsl:apply\-templates\ select="\@*|node()"/>
</xsl:copy>
</xsl:template>
\f[]
.fi
.PP
This means any elements or attributes not matched by a more specific
template in the script are copied.
.SS Creating presentation applicability statements
.PP
Given the following:
.IP
.nf
\f[C]
<dmodule>
<identAndStatusSection>
<dmAddress>...</dmAddress>
<dmStatus>
\&...
<applic>
<displayText>
<simplePara>A\ or\ B</simplePara>
</displayText>
</applic>
\&...
</dmStatus>
</identAndStatusSection>
<content>
<referencedApplicGroup>
<applic\ id="applic\-B">
<displayText>
<simplePara>B</simplePara>
</displayText>
</applic>
</referencedApplicGroup>
<procedure>
<preliminaryRqmts>...</preliminaryRqmts>
<mainProcedure>
<proceduralStep>
<para>This\ step\ is\ applicable\ to\ A\ or\ B.</para>
</proceduralStep>
<proceduralStep\ applicRefId="applic\-B">
<para>This\ step\ is\ applicable\ to\ B\ only.</para>
</proceduralStep>
<proceduralStep\ applicRefId="applic\-B">
<para>This\ step\ is\ also\ applicable\ to\ B\ only.</para>
</proceduralStep>
<proceduralStep>
<para>This\ step\ is\ also\ applicable\ to\ A\ or\ B.</para>
</proceduralStep>
</mainProcedure>
<closeRqmts>...</closeRqmts>
</procedure>
</content>
</dmodule>
\f[]
.fi
.PP
Applicability statements should be displayed whenever applicability
changes:
.IP "1." 3
This step is applicable to A or B.
.IP "2." 3
\f[I]Applicable to: B\f[]
.RS 4
.PP
This step is applicable to B only.
.RE
.IP "3." 3
This step is also applicable to B only.
.IP "4." 3
\f[I]Applicable to: A or B\f[]
.RS 4
.PP
This step is also applicable to A or B.
.RE
.PP
There are two parts which are difficult to do in an XSLT stylesheet:
.IP \[bu] 2
No statement is shown on Step 3 despite having attribute
\f[C]applicRefId\f[] because the applicability has not changed since the
last statement on Step 2.
.IP \[bu] 2
A statement is shown on Step 4 despite not having attribute
\f[C]applicRefId\f[] because the applicability has changed back to that
of the whole data module.
.PP
Using the s1kd\-aspp tool, the above XML would produce the following
output:
.IP
.nf
\f[C]
<dmodule>
<identAndStatusSection>
<dmAddress>...</dmAddress>
<dmStatus>
\&...
<applic>
<displayText>
<simplePara>A\ or\ B</simplePara>
</displayText>
</applic>
\&...
</dmStatus>
</identAndStatusSection>
<content>
<referencedApplicGroup>
<applic\ id="applic\-B">
<displayText>
<simplePara>B</simplePara>
</displayText>
</applic>
<applic\ id="applic\-0000">
<displayText>
<simplePara>A\ or\ B</simplePara>
</displayText>
</applic>
</referencedApplicGroup>
<procedure>
<preliminaryRqmts>...</preliminaryRqmts>
<mainProcedure>
<proceduralStep>
<para>This\ step\ is\ applicable\ to\ A\ or\ B.</para>
</proceduralStep>
<proceduralStep\ applicRefId="applic\-B">
<para>This\ step\ is\ applicable\ to\ B\ only.</para>
</proceduralStep>
<proceduralStep>
<para>This\ step\ is\ also\ applicable\ to\ B\ only.</para>
</proceduralStep>
<proceduralStep\ applicRefId="applic\-0000">
<para>This\ step\ is\ also\ applicable\ to\ A\ or\ B.</para>
</proceduralStep>
</mainProcedure>
</procedure>
</content>
</dmodule>
\f[]
.fi
.PP
With attribute \f[C]applicRefId\f[] only on those elements where a
statement should be shown, and an additional inline applicability to
represent the whole data module\[aq]s applicability.
This XML is semantically incorrect but easier for a stylesheet to
transform for page\-oriented output.
.SH AUTHORS
khzae.net.


s1kd-aspp/s1kd-aspp.c
/* Applicability Preprocessor
 *
 * Preprocesses a data module's applicability statements (@applicRefId) in to a
 * simpler format which is easier for an XSL stylesheet to process.
 *
 * The applicability in the resulting output will not be semantically correct. */

#include <stdio.h>
#include <unistd.h>
#include <stdbool.h>
#include <string.h>
#include <libxml/tree.h>
#include <libxml/xpath.h>
#include <libxml/xinclude.h>
#include <libxslt/transform.h>
#include <libexslt/exslt.h>
#include "elements_list.h"
#include "generateDisplayText.h"
#include "identity.h"

/* ID for the inline <applic> element representing the whole data module's
 * applicability. */
#define DEFAULT_DM_APPLIC_ID BAD_CAST "applic-0000"
xmlChar *dmApplicId;

/* XPath to select all elements which may have applicability annotations.
 *
 * Read from elements_list.h*/
xmlChar *applicElemsXPath;

/* Bug in libxml < 2.9.2 where parameter entities are resolved even when
 * XML_PARSE_NOENT is not specified.
 */
#if LIBXML_VERSION < 20902
#define PARSE_OPTS XML_PARSE_NONET
#else
#define PARSE_OPTS 0
#endif

char *customGenDispTextXsl = NULL;

/* Return the first node matching an XPath expression. */
xmlNodePtr firstXPathNode(xmlDocPtr doc, xmlNodePtr node, const char *xpath)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlNodePtr first;

	if (doc) {
		ctx = xmlXPathNewContext(doc);
	} else if (node) {
		ctx = xmlXPathNewContext(node->doc);
	} else {
		return NULL;
	}

	ctx->node = node;

	obj = xmlXPathEvalExpression(BAD_CAST xpath, ctx);

	if (xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		first = NULL;
	} else {
		first = obj->nodesetval->nodeTab[0];
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);

	return first;
}

/* Return the last node matching an XPath expression. */
xmlNodePtr lastXPathNode(xmlDocPtr doc, xmlNodePtr node, const char *xpath)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlNodePtr last;

	if (doc) {
		ctx = xmlXPathNewContext(doc);
	} else if (node) {
		ctx = xmlXPathNewContext(node->doc);
	} else {
		return NULL;
	}

	ctx->node = node;

	obj = xmlXPathEvalExpression(BAD_CAST xpath, ctx);

	if (xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		last = NULL;
	} else {
		last = obj->nodesetval->nodeTab[obj->nodesetval->nodeNr - 1];
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);

	return last;
}

/* Set an explicit applicability statement on a node.
 *
 * Nodes with the attribute @applicRefId already have an explicit statement.
 *
 * Nodes without @applicRefId have an implicit statement inherited from their
 * parent/ancestor which does have an explicit statement, or ultimately from
 * the applicability of the whole data module. */
void processNode(xmlNodePtr node)
{
	xmlNodePtr attr;

	attr = firstXPathNode(NULL, node, "@applicRefId|@refapplic");

	if (!attr) {
		/* Inherit applicability from
		 * - an ancestor, or
		 * - the whole data module level */
		xmlNodePtr ancestor;
		xmlChar *name;

		ancestor = lastXPathNode(NULL, node, "ancestor::*[@applicRefId]|ancestor::*[@refapplic]");

		name = BAD_CAST (firstXPathNode(NULL, node, "//idstatus") ? "refapplic" : "applicRefId");

		if (!ancestor) {
			xmlSetProp(node, name, dmApplicId);
		} else {
			xmlChar *ancestorApplic = xmlGetProp(ancestor, name);
			xmlSetProp(node, name, ancestorApplic);
			xmlFree(ancestorApplic);
		}
	}
}

/* Set explicit applicability on all nodes in a nodeset. */
void processNodeSet(xmlNodeSetPtr nodes)
{
	int i;

	for (i = 0; i < nodes->nodeNr; ++i) {
		processNode(nodes->nodeTab[i]);
	}
}

/* Remove duplicate applicability statements in document-order so that a
 * statement is only shown when applicability changes. */
void removeDuplicates(xmlNodeSetPtr nodes)
{
	int i;

	xmlChar *applic = xmlStrdup(dmApplicId);

	for (i = 0; i < nodes->nodeNr; ++i) {
		xmlNodePtr attr;
		xmlChar *applicRefId;

		attr = firstXPathNode(NULL, nodes->nodeTab[i], "@applicRefId|@refapplic");
		applicRefId = xmlNodeGetContent(attr);

		if (xmlStrcmp(applicRefId, applic) == 0) {
			xmlUnsetProp(nodes->nodeTab[i], attr->name);
		} else {
			xmlFree(applic);
			applic = xmlStrdup(applicRefId);
		}

		xmlFree(applicRefId);
	}

	xmlFree(applic);
}

/* Insert a new inline <applic> element representing the applicability of the
 * whole data module, based on the dmStatus/applic element. */
void addDmApplic(xmlNodePtr dmodule)
{
	xmlNodePtr referencedApplicGroup;

	if ((referencedApplicGroup = firstXPathNode(NULL, dmodule, ".//referencedApplicGroup|.//inlineapplics"))) {
		xmlNodePtr applic;
		xmlNodePtr wholeDmApplic;

		wholeDmApplic = firstXPathNode(NULL, dmodule, ".//dmStatus/applic|.//status/applic");
		applic = xmlAddChild(referencedApplicGroup, xmlCopyNode(wholeDmApplic, 1));

		xmlSetProp(applic, BAD_CAST "id", dmApplicId);
	}
}

void dumpGenDispTextXsl(void)
{
	printf("%.*s", generateDisplayText_xsl_len, generateDisplayText_xsl);
}

void addIdentity(xmlDocPtr style)
{
	xmlDocPtr identity;
	xmlNodePtr stylesheet, first, template;

	identity = xmlReadMemory((const char *) ___common_identity_xsl, ___common_identity_xsl_len, NULL, NULL, 0);
	template = xmlFirstElementChild(xmlDocGetRootElement(identity));

	stylesheet = xmlDocGetRootElement(style);

	first = xmlFirstElementChild(stylesheet);

	if (first) {
		xmlAddPrevSibling(first, xmlCopyNode(template, 1));
	} else {
		xmlAddChild(stylesheet, xmlCopyNode(template, 1));
	}

	xmlFreeDoc(identity);
}

void generateDisplayText(xmlDocPtr doc, xmlNodePtr acts, xmlNodePtr ccts)
{
	xmlDocPtr styledoc, res, muxdoc;
	xsltStylesheetPtr style;
	xmlNodePtr mux, cur, muxacts, muxccts, new, old;

	muxdoc = xmlNewDoc(BAD_CAST "1.0");
	mux = xmlNewNode(NULL, BAD_CAST "mux");
	xmlDocSetRootElement(muxdoc, mux);

	xmlAddChild(mux, xmlCopyNode(xmlDocGetRootElement(doc), 1));
	muxacts = xmlNewChild(mux, NULL, BAD_CAST "acts", NULL);
	muxccts = xmlNewChild(mux, NULL, BAD_CAST "ccts", NULL);
	for (cur = acts->children; cur; cur = cur->next) {
		xmlDocPtr act;
		xmlChar *path;
		path = xmlNodeGetContent(cur);
		act = xmlReadFile((char *) path, NULL, PARSE_OPTS);
		xmlAddChild(muxacts, xmlCopyNode(xmlDocGetRootElement(act), 1));
		xmlFreeDoc(act);
		xmlFree(path);
	}
	for (cur = ccts->children; cur; cur = cur->next) {
		xmlDocPtr cct;
		xmlChar *path;
		path = xmlNodeGetContent(cur);
		cct = xmlReadFile((char *) path, NULL, PARSE_OPTS);
		xmlAddChild(muxccts, xmlCopyNode(xmlDocGetRootElement(cct), 1));
		xmlFreeDoc(cct);
		xmlFree(path);
	}

	if (customGenDispTextXsl) {
		styledoc = xmlReadFile(customGenDispTextXsl, NULL, PARSE_OPTS);
	} else {
		styledoc = xmlReadMemory((const char *) generateDisplayText_xsl,
			generateDisplayText_xsl_len, NULL, NULL, 0);
	}

	addIdentity(styledoc);

	style = xsltParseStylesheetDoc(styledoc);

	res = xsltApplyStylesheet(style, muxdoc, NULL);

	new = xmlCopyNode(firstXPathNode(res, NULL, "/mux/dmodule"), 1);
	old = xmlDocSetRootElement(doc, new);
	xmlFreeNode(old);

	xmlFreeDoc(res);
	xmlFreeDoc(muxdoc);
	xsltFreeStylesheet(style);
}

void processDmodule(xmlNodePtr dmodule)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;

	ctx = xmlXPathNewContext(dmodule->doc);
	ctx->node = dmodule;
	obj = xmlXPathEvalExpression(applicElemsXPath, ctx);

	processNodeSet(obj->nodesetval);
	removeDuplicates(obj->nodesetval);
	addDmApplic(dmodule);

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

void processDmodules(xmlNodeSetPtr dmodules)
{
	int i;

	for (i = 0; i < dmodules->nodeNr; ++i) {
		processDmodule(dmodules->nodeTab[i]);
	}
}

void processFile(const char *in, const char *out, bool xincl, bool process,
	bool genDispText, xmlNodePtr acts, xmlNodePtr ccts)
{
	xmlDocPtr doc;
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;

	doc = xmlReadFile(in, NULL, PARSE_OPTS);

	if (xincl) {
		xmlXIncludeProcess(doc);
	}

	if (process) {
		ctx = xmlXPathNewContext(doc);
		obj = xmlXPathEvalExpression(BAD_CAST "//dmodule", ctx);

		processDmodules(obj->nodesetval);

		xmlXPathFreeObject(obj);
		xmlXPathFreeContext(ctx);
	}

	if (genDispText) {
		generateDisplayText(doc, acts, ccts);
	}

	xmlSaveFile(out, doc);

	xmlFreeDoc(doc);
}

void showHelp(void)
{
	puts("Usage: s1kd-aspp [-g [-A <ACT>] [-C <CCT>]] [-p [-a <ID>]] [-dfxh?] [<modules>]");
	puts("");
	puts("Options:");
	puts("  -A <ACT>  Use <ACT> when generating display text.");
	puts("  -a <ID>   Use <ID> for DM-level applic.");
	puts("  -C <CCT>  Use <CCT> when generating display text.");
	puts("  -d        Dump built-in XSLT for generating display text.");
	puts("  -f        Overwrite input file(s).");
	puts("  -G <XSL>  Use custom XSLT script to generate display text.");
	puts("  -g        Generate display text for applicability statements.");
	puts("  -p        Convert semantic applicability to presentation applicability.");
	puts("  -h -?     Show help/usage message.");
}

int main(int argc, char **argv)
{
	int i;
	bool overwrite = false;
	bool xincl = false;
	bool genDispText = false;
	bool process = false;
	
	xmlNodePtr acts, ccts;

	exsltRegisterAll();

	dmApplicId = xmlStrdup(DEFAULT_DM_APPLIC_ID);

	applicElemsXPath = xmlStrndup(elements_list, elements_list_len);

	acts = xmlNewNode(NULL, BAD_CAST "acts");
	ccts = xmlNewNode(NULL, BAD_CAST "ccts");

	while ((i = getopt(argc, argv, "A:a:C:dfG:gpxh?")) != -1) {
		switch (i) {
			case 'A':
				xmlNewChild(acts, NULL, BAD_CAST "act", BAD_CAST optarg);
				break;
			case 'a':
				xmlFree(dmApplicId);
				dmApplicId = xmlStrdup(BAD_CAST optarg);
				break;
			case 'C':
				xmlNewChild(ccts, NULL, BAD_CAST "cct", BAD_CAST optarg);
				break;
			case 'd':
				dumpGenDispTextXsl();
				exit(0);
			case 'f':
				overwrite = true;
				break;
			case 'G':
				customGenDispTextXsl = strdup(optarg);
				break;
			case 'g':
				genDispText = true;
				break;
			case 'p':
				process = true;
				break;
			case 'x':
				xincl = true;
				break;
			case 'h':
			case '?':
				showHelp();
				exit(0);
		}
	}

	if (optind >= argc) {
		processFile("-", "-", xincl, process, genDispText, acts, ccts);
	} else {
		for (i = optind; i < argc; ++i) {
			processFile(argv[i], overwrite ? argv[i] : "-", xincl,
				process, genDispText, acts, ccts);
		}
	}

	xmlFree(dmApplicId);
	xmlFree(applicElemsXPath);

	xmlFreeNode(acts);
	xmlFreeNode(ccts);

	free(customGenDispTextXsl);

	xsltCleanupGlobals();
	xmlCleanupParser();

	return 0;
}
